diff --git a/KernelSU/kernel/core_hook.c b/KernelSU/kernel/core_hook.c
index c6ac7d66..4eb417eb 100644
--- a/KernelSU/kernel/core_hook.c
+++ b/KernelSU/kernel/core_hook.c
@@ -284,7 +284,7 @@ int ksu_handle_prctl(int option, unsigned long arg2, unsigned long arg3,
 
 	if (arg2 == CMD_GRANT_ROOT) {
 		if (is_allow_su()) {
-			pr_info("allow root for: %d\n", current_uid());
+			pr_info("allow root for: %d\n", current_uid().val);
 			escape_to_root();
 			if (copy_to_user(result, &reply_ok, sizeof(reply_ok))) {
 				pr_err("grant_root: prctl reply error\n");
@@ -298,7 +298,7 @@ int ksu_handle_prctl(int option, unsigned long arg2, unsigned long arg3,
 		if (is_manager() || 0 == current_uid().val) {
 			u32 version = KERNEL_SU_VERSION;
 			if (copy_to_user(arg3, &version, sizeof(version))) {
-				pr_err("prctl reply error, cmd: %d\n", arg2);
+				pr_err("prctl reply error, cmd: %lu\n", arg2);
 			}
 		}
 		return 0;
@@ -372,7 +372,7 @@ int ksu_handle_prctl(int option, unsigned long arg2, unsigned long arg3,
 						  sizeof(u32) * array_length)) {
 					if (copy_to_user(result, &reply_ok,
 							 sizeof(reply_ok))) {
-						pr_err("prctl reply error, cmd: %d\n",
+						pr_err("prctl reply error, cmd: %lu\n",
 						       arg2);
 					}
 				} else {
@@ -392,16 +392,16 @@ int ksu_handle_prctl(int option, unsigned long arg2, unsigned long arg3,
 			} else if (arg2 == CMD_UID_SHOULD_UMOUNT) {
 				allow = ksu_uid_should_umount(target_uid);
 			} else {
-				pr_err("unknown cmd: %d\n", arg2);
+				pr_err("unknown cmd: %lu\n", arg2);
 			}
 			if (!copy_to_user(arg4, &allow, sizeof(allow))) {
 				if (copy_to_user(result, &reply_ok,
 						 sizeof(reply_ok))) {
-					pr_err("prctl reply error, cmd: %d\n",
+					pr_err("prctl reply error, cmd: %lu\n",
 					       arg2);
 				}
 			} else {
-				pr_err("prctl copy err, cmd: %d\n", arg2);
+				pr_err("prctl copy err, cmd: %lu\n", arg2);
 			}
 		}
 		return 0;
@@ -428,7 +428,7 @@ int ksu_handle_prctl(int option, unsigned long arg2, unsigned long arg3,
 				return 0;
 			}
 			if (copy_to_user(result, &reply_ok, sizeof(reply_ok))) {
-				pr_err("prctl reply error, cmd: %d\n", arg2);
+				pr_err("prctl reply error, cmd: %lu\n", arg2);
 			}
 		}
 		return 0;
@@ -444,7 +444,7 @@ int ksu_handle_prctl(int option, unsigned long arg2, unsigned long arg3,
 		// todo: validate the params
 		if (ksu_set_app_profile(&profile, true)) {
 			if (copy_to_user(result, &reply_ok, sizeof(reply_ok))) {
-				pr_err("prctl reply error, cmd: %d\n", arg2);
+				pr_err("prctl reply error, cmd: %lu\n", arg2);
 			}
 		}
 		return 0;
diff --git a/KernelSU/kernel/ksud.c b/KernelSU/kernel/ksud.c
index 55cd1997..b3791884 100644
--- a/KernelSU/kernel/ksud.c
+++ b/KernelSU/kernel/ksud.c
@@ -283,17 +283,17 @@ int ksu_handle_vfs_read(struct file **file_ptr, char __user **buf_ptr,
 
 	size_t rc_count = strlen(KERNEL_SU_RC);
 
-	pr_info("vfs_read: %s, comm: %s, count: %d, rc_count: %d\n", dpath,
+	pr_info("vfs_read: %s, comm: %s, count: %lu, rc_count: %lu\n", dpath,
 		current->comm, count, rc_count);
 
 	if (count < rc_count) {
-		pr_err("count: %d < rc_count: %d", count, rc_count);
+		pr_err("count: %lu < rc_count: %lu", count, rc_count);
 		return 0;
 	}
 
 	size_t ret = copy_to_user(buf, KERNEL_SU_RC, rc_count);
 	if (ret) {
-		pr_err("copy ksud.rc failed: %d\n", ret);
+		pr_err("copy ksud.rc failed: %lu\n", ret);
 		return 0;
 	}
 
diff --git a/KernelSU/kernel/uid_observer.c b/KernelSU/kernel/uid_observer.c
index b12eff11..022f3778 100644
--- a/KernelSU/kernel/uid_observer.c
+++ b/KernelSU/kernel/uid_observer.c
@@ -42,7 +42,7 @@ static void do_update_uid(struct work_struct *work)
 	struct file *fp = ksu_filp_open_compat(SYSTEM_PACKAGES_LIST_PATH, O_RDONLY, 0);
 	if (IS_ERR(fp)) {
 		pr_err("do_update_uid, open " SYSTEM_PACKAGES_LIST_PATH
-		       " failed: %d\n",
+		       " failed: %ld\n",
 		       PTR_ERR(fp));
 		return;
 	}
